File: Scrabble-v2/mobile/lib/services/game_service.dart
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:flutter_translate/flutter_translate.dart';
import 'package:poly_scrabble/models/user.dart';
import 'package:poly_scrabble/services/popup_service.dart';

import '../../models/player.dart';
import '../constants/board_constants.dart';
import '../helpers/helpers.dart';
import '../models/board.dart';
import '../services/placementString_service.dart';
import 'exchangeString_service.dart';
import 'fetchBoard_service.dart';
import 'socket_service.dart';

class GameService extends ChangeNotifier {
  final SocketService socketService;
  final UserModel userModel;
  String currentPlayer = '';
  int _remainingTime = 0;
  int remainingLetters = 0;

  final Map<int, Tile?> easel = {};
  /*{0: Tile.isJoker(
        Position(100, 150), '', 8, TileState.easel, Coordinate(0, 0), true),
    1: Tile.isJoker(
        Position(250, 250), '', 1, TileState.easel, Coordinate(0, 0), true),
    2: Tile(Position(250, 250), 'B', 1, TileState.easel, Coordinate(0, 0)),
    3: Tile(Position(250, 250), 'C', 1, TileState.easel, Coordinate(0, 0)),
    4: Tile(Position(250, 250), 'D', 1, TileState.easel, Coordinate(0, 0)),
    5: Tile(Position(250, 250), 'E', 1, TileState.easel, Coordinate(0, 0)),
    6: Tile(Position(250, 250), 'F', 1, TileState.easel, Coordinate(0, 0)),
  };*/

  final placedTiles = List<List<Tile?>>.generate(
      numberOfSquaresInBoard.toInt() + 1,
      (index) => List<Tile?>.generate(
          numberOfSquaresInBoard.toInt() + 1, (index) => null,
          growable: false),
      growable: false);

  bool isGameOver = false;

  Coordinate? placementCursor;

  String remainingTime() {
    return formatTime(_remainingTime);
  }

  GameService(this.socketService, this.userModel) {
    setSocketListeners();
  }

  void resetPlacedTiles() {
    for (var i = 0; i < placedTiles.length; i++) {
      for (var j = 0; j < placedTiles[i].length; j++) {
        placedTiles[i][j] = null;
      }
    }
  }

  void resetEasel() {
    easel.forEach((key, value) {
      easel[key] = null;
    });
  }

  void reset() {
    currentPlayer = '';
    _remainingTime = 0;
    remainingLetters = 0;
    resetEasel();
    resetPlacedTiles();
    notifyListeners();
  }

  bool isConnectedPlayerTurn() {
    return currentPlayer == userModel.username;
  }

  void setSocketListeners() {
    socketService.on('playerTurnChanged', (data) {
      try {
        placementCursor = null;
        currentPlayer = data;
        if (!isConnectedPlayerTurn()) {
          returnTilesToEasel();
        }
        notifyListeners();
      } catch (e) {
        log("Error while changing turn ${e.toString()}");
      }
    });
    socketService.on('timeUpdated', (data) {
      try {
        _remainingTime = data;
        notifyListeners();
      } catch (e) {
        log("Error while updating time, ${e.toString()}");
      }
    });
    socketService.on('lettersBankCountUpdated', (data) {
      try {
        remainingLetters = data;
        notifyListeners();
      } catch (e) {
        log("Error while updating letter count, ${e.toString()}");
      }
    });

    socketService.on('drawBoard', (data) {
      try {
        PlacementStringService.updatePlacedTiles(placedTiles, data);
        notifyListeners();
      } catch (e) {
        log("Error while updating draw board, $e");
      }
    });
    socketService.on('drawRack', (data) {
      try {
        ExchangeStringService().updateEasel(easel, data.trim());
        notifyListeners();
      } catch (e) {
        log("Error while updating draw rack, ${e.toString()}");
      }
    });
    socketService.on('drawBoardCursor', (data) {
      try {
        if(data['x']==-1){ placementCursor = null; }
        else{ placementCursor = Coordinate(data['x'], data['y']); }
        notifyListeners();
      } catch (e) {
        log("Error while updating draw placement cursor, ${e.toString()}");
      }
    });

    socketService.on('fetchBoard', (data) {
      try {
        FetchBoardService.initializePlacedTiles(placedTiles, data);
        notifyListeners();
      } catch (e) {
        log("Error while fetching board, ${e.toString()}");
      }
    });

    socketService.on('gameIsOver', (dynamic winnerArray) {
      try {
        log('receiving "gameIsOver" event');
        isGameOver = true;
        List<Player> players = [];
        Player tempPlayer;
        var endGameText = translate("end_game.winners");
        for (var winner in winnerArray) {
          tempPlayer = Player();
          tempPlayer.pseudo = winner['pseudo'] ?? '';
          tempPlayer.points = winner['points'] ?? 0;
          players.add(tempPlayer);
          endGameText += '${tempPlayer.pseudo} ';
        }
        PopupService.hey(endGameText);
        notifyListeners();
      } catch (e) {
        log("Error while changing gameIsOver, ${e.toString()}");
      }
    });
  }

  void sendPlacementCursor(String roomName, Coordinate? placementCursor) {
    this.placementCursor = placementCursor;
    if (placementCursor != null) {
      dynamic data = {
        'roomName': roomName,
        'x': placementCursor?.x,
        'y': placementCursor?.y
      };
      socketService.send('drawBoardCursor', data);
    }
    else{
      dynamic data = {
        'roomName': roomName,
        'x': -1,
        'y': -1
      };
      socketService.send('drawBoardCursor', data);
    }
  }

  void sendPlacement() {
    socketService.send('command',
        PlacementStringService().placementCommand(easel, placedTiles));
  }

  void sendExchange() {
    socketService.send(
        'command', ExchangeStringService().exchangeCommand(easel));
  }

  void sendPass() {
    socketService.send('command', '!passer');
  }

  void returnTilesToEasel() {
    for (int index in easel.keys) {
      if (easel[index] != null) {
        easel[index]?.position = null;
        easel[index]?.state = TileState.easel;
        if(easel[index]?.isJoker == true){
          easel[index]?.letter = '*';
        }
      }
    }
  }
}

File: Scrabble-v2/web/src/app/services/game-service.ts
import { Injectable } from '@angular/core';
import { ArrowCursor } from '@app/classes/board/arrow-cursor';
import { Coordinate } from '@app/classes/board/coordinate';
import { ExchangeStringService } from '@app/classes/board/exchange';
import { FetchBoard } from '@app/classes/board/fetch-board';
import { Letter } from '@app/classes/board/letter';
import { PlacementString } from '@app/classes/board/placement';
import { Player } from '@app/classes/player';
import { DEFAULT_CASE_COUNT } from '@app/constants/board-constants';
import { LetterState } from '@app/enums/letter-state';
import { BehaviorSubject } from 'rxjs';
import { SocketClientService } from './socket-client.service';

@Injectable({
    providedIn: 'root',
})
export class GameService {
    easel: BehaviorSubject<(Letter | undefined)[]> = new BehaviorSubject<(Letter | undefined)[]>([]);
    arrowCursor: BehaviorSubject<ArrowCursor | undefined>;
    placementCursor: BehaviorSubject<Coordinate | undefined> = new BehaviorSubject<Coordinate | undefined>(undefined);
    placedLetters: BehaviorSubject<(Letter | undefined)[][]> = new BehaviorSubject<(Letter | undefined)[][]>([]);
    isConnectedPlayerTurn: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);

    constructor(private socketClientService: SocketClientService, public player: Player) {
        for (let i = 0; i < DEFAULT_CASE_COUNT; i++) {
            const row = new Array(DEFAULT_CASE_COUNT).fill(undefined);
            this.placedLetters.value.push(row);
        }
        this.easel.next([
            { position: undefined, char: 'a', value: 1, state: LetterState.Easel, coordinate: undefined, isJoker: false } as Letter,
            { position: undefined, char: '*', value: 0, state: LetterState.Easel, coordinate: undefined, isJoker: true } as Letter,
            { position: undefined, char: 'c', value: 1, state: LetterState.Easel, coordinate: undefined, isJoker: false } as Letter,
            { position: undefined, char: 'd', value: 1, state: LetterState.Easel, coordinate: undefined, isJoker: false } as Letter,
            { position: undefined, char: 'e', value: 1, state: LetterState.Easel, coordinate: undefined, isJoker: false } as Letter,
            { position: undefined, char: 'z', value: 10, state: LetterState.Easel, coordinate: undefined, isJoker: false } as Letter,
            { position: undefined, char: 'k', value: 10, state: LetterState.Easel, coordinate: undefined, isJoker: false } as Letter,
        ]);
        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
        this.arrowCursor = new BehaviorSubject<ArrowCursor | undefined>(undefined);
        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    }
    initSocket() {
        this.socketClientService.on('drawRack', (data: string) => {
            ExchangeStringService.updateEasel(this.easel.value, data.trim());
            this.easel.next(this.easel.value);
        });
        this.socketClientService.on('drawBoard', (data: unknown) => {
            PlacementString.updatePlacedLetters(this.placedLetters.value, data);
            this.placedLetters.next(this.placedLetters.value);
        });
        this.socketClientService.on('drawBoardCursor', (data: Coordinate) => {
            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
            if (data.x === -1) this.placementCursor.next(undefined);
            else this.placementCursor.next(new Coordinate(data.x, data.y));
        });
        this.socketClientService.on('fetchBoard', (data: string[]) => {
            FetchBoard.initializePlacedLetters(this.placedLetters.value, data);
            this.placedLetters.next(this.placedLetters.value);
        });

        this.socketClientService.on('playerTurnChanged', (data: string) => {
            this.isConnectedPlayerTurn.next(data === this.player.pseudo);
            this.arrowCursor.next(undefined);
            this.placementCursor.next(undefined);
        });
    }

    resetPlacedLetters() {
        this.placedLetters = new BehaviorSubject<(Letter | undefined)[][]>([]);
        for (let i = 0; i < DEFAULT_CASE_COUNT; i++) {
            const row = new Array(DEFAULT_CASE_COUNT).fill(undefined);
            this.placedLetters.value.push(row);
        }
        this.placedLetters.next(this.placedLetters.value);
    }

    reset() {
        this.resetPlacedLetters();
        this.placementCursor.next(undefined);
        this.arrowCursor.next(undefined);
        this.easel.next([]);
        this.isConnectedPlayerTurn.next(false);
    }

    sendPlacementCursor(room: string) {
        let data;
        if (this.placementCursor.value) {
            data = {
                roomName: room,
                x: this.placementCursor.value.x,
                y: this.placementCursor.value.y,
            };
        } else {
            data = {
                roomName: room,
                x: -1,
                y: -1,
            };
        }
        this.socketClientService.send('drawBoardCursor', data);
    }
    sendExchange() {
        this.socketClientService.send('command', ExchangeStringService.exchangeCommand(this.easel.value)); // ajouter d'envoyer le string
    }
    sendSkip() {
        this.socketClientService.send('command', '!passer');
    }
    sendPlay() {
        const placementCommand = PlacementString.placementCommand(this.easel.value, this.placedLetters.value);
        this.socketClientService.send('command', placementCommand);
    }
}
